### PR TITLE
Fix Chunked keepalives

### LINKED ISSUE #598: Keep-alive broken in 1.10.3 with chunked transfer-encoding
Per kennethreitz/requests#2568, starting in urllib3 1.10.3 an exception is thrown if a connection with chunked transfer-encoding is reused.

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/11976603-keep-alive-broken-in-1-10-3-with-chunked-transfer-encoding?utm_campaign=plugin&utm_content=tracker%2F192525&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F192525&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by t-8ch**
I am not sure if this is the right way. The test should be picked anyways.

Fixes #598.

/cc @shazow @sigmavirus24

### CODE REVIEW

File: dummyserver/handlers.py [hunk 1]
```diff
@@ -34,7 +31,20 @@ def __call__(self, request_handler):
...
+                    item = item.encode('utf8')
+            request_handler.write(item)
+            request_handler.flush()
+        else:
+            body = self.body
+            if not isinstance(body, bytes):
+                body = body.encode('utf8')
+
+            request_handler.write(body)
+
```
Comment (Lukasa) [inline dummyserver/handlers.py]: What's all this for? As far as I know, chunked responses never return a list as the body.

**Comment by Lukasa** [comment]
This looks like it fixes the core issue, but there seems to be some unrelated work in here. Let's sort that out before merging. =)

File: dummyserver/handlers.py (hunk 1 above)
Comment (t-8ch) [inline dummyserver/handlers.py]: This is meant to make the `chunked()`-endpoint work

Comment (Lukasa) [inline dummyserver/handlers.py]: Ah, fair enough. I take it you confirmed that this genuinely does write correctly-framed chunked transfer encoding? (As in, the request_handler correctly applies the chunked framing)

Comment (t-8ch) [inline dummyserver/handlers.py]: At least in triggers our bug :-)
(And I reconfirmed it on my side)

Comment (t-8ch) [inline dummyserver/handlers.py]: Woah, wait. It seems to discard output (but is chunked...) I have to check again...

Comment (t-8ch) [inline dummyserver/handlers.py]: I messed up during final cleanup. It would only emit the last chunk. Fixed now

Comment (Lukasa) [inline dummyserver/handlers.py]: On the one hand, I'm glad I asked! On the other hand, now the tests failed. =(

**Comment by Lukasa** [comment]
Tests failing, otherwise I'm happy. =)

**Comment by t-8ch** [comment]
Sometimes we have resource warnings in py3....

**Comment by Lukasa** [comment]
Dem intermittent test failures.

The resource warnings aren't *necessarily* a surprise (they happen when used connection pools go out of scope) but if they're intermittent then they are a surprise.

File: test/with_dummyserver/test_connectionpool.py [hunk 2]
```diff
@@ -618,6 +618,25 @@ def test_source_address_error(self):
...
+
+        for _ in range(x):
+            response = self.pool.request(
+                    'GET',
+                    '/chunked',
+                    headers={
+                        'Connection': 'keep-alive',
+                        },
+                    preload_content=False,
+                    retries=0,
```
Comment (shazow) [inline test/with_dummyserver/test_connectionpool.py]: `retries=False` disables retries (and exception wrapping).

Comment (t-8ch) [inline test/with_dummyserver/test_connectionpool.py]: @shazow changed

**Comment by shazow** [comment]
Looks alright to me, lmk when you deem it done. :)

**Comment by t-8ch** [comment]
@shazow If you like it, it is done :-)

**Pull request merged**